README.md
# B-Tree Implementation
Implementation of a B-Tree data structure for the Programming Foundations II class at the University of Arkansas

Instructor: Joshua Burbridge

Contributors: Davis Campbell, Chris Carter, Wesley Nimmo, and Sam Yeager

# How to compile and run:

Option 1: Codeblocks

1. Clone/download the main.cpp file and the two .h files to your computer using a method of your choice

2. Create a new console application c++ project in codeblocks

3. Open the three files you downloaded/cloned in codeblocks

4. Right click on the tabs for each of the files in codeblocks and click "add to project"

5. Click on the compile and run icon

Option 2: Unix based computers (with g++ installed)

1. Open your terminal and navigate to a directory in which you would like to store the files

2. Type the command "git clone https://github.com/cicarter/homework5"

3. Type the command "g++ main.cpp -o run.exe"

4. Type the command "./run.exe" to run the program


changelog.txt
07/28/18 - Davis
- Added remove helper method
- fixed various bugs and errors
- moved main.cpp to separate file
- condensed the .cpp and .h for both classes into one file each as part of bug fixing
- re-named the "M" variable in the node class to order, because that's what it was
- added isLeaf method
- improved logic of isFull and isEmpty
- added changelog.txt file
- added "keys" variable, which keeps track of the number of keys present in a node

07/29/18 - Davis - commit 1
- added friend class line to node class. setters and getters no longer needed.
- Only commented out the setter and getter related changes so you can see what i've done. we can delete later
- did a bunch of work on the remove method (just by following the logic i found online it's gonna be long af)

07/29/18 - Davis commit 2
- Finished both remove methods (i was right. they're long af)
WARNING: to whoever is doing the insert method, it will also be very long, so watch out

07/29/18 - Davis commit 3
- changed the word "order" with the word "degree" in all cases.

07/30/18 - Davis
- fixed a few places that were messed up when changing from going by order to by degree
- fixed a few places that were messed up when removing the getters/setters
- got a temporary insert implementation from the internet to test my remove method. DONT FORGET TO REMOVE lol
- finally deleted setters/getters
- also added print function from internet for temporary use ALSO DON'T FORGET TO REMOVE
- fixed all bugs in my remove method (took a long time haha)
- updated main for better testing

08/01/18 - Davis commit 1
- No one else used the changelog lol
- removed both insert functions and wrote my own
- removed both print functions and wrote my own

08/01/18 - Davis commit 2
- fixed all the things
- cleaned up everything



main.cpp
#include "b_tree.h"
#include <iostream>
#include <time.h>

using namespace std;
int main()
{
    /*
    No error checking on inputs for this, so user must be skilled lol
    */
    srand(time(NULL));
    int size; // this number can be literally as big as you want it to be haha
    int choice, value, degree, type, length;
    string word;
    char character;
    cout << "Define your B-Tree\n"
         << "Degree: ";
    cin >> degree;
    cout << "Choose data type:\n"
         << "(1) Integers\n"
         << "(2) Characters\n"
         << "(3) Strings\n"
         << "Your choice: ";
    cin >> type;

    BTree<int> ints(degree);
    BTree<char> chars(degree);
    BTree<string> strings(degree);

    while (choice != 6)
    {
        cout << "\nMenu:\n"
             << "(1) Insert a data point\n"
             << "(2) Search for a value\n"
             << "(3) Insert a bunch of random values\n"
             << "(4) Remove a value\n"
             << "(5) Print the tree\n"
             << "(6) Quit\n"
             << "Your choice: ";
        cin >> choice;

        if (choice == 1)
        {
            if (type == 1)
            {
                cout << "Number: ";
                cin >> value;
                ints.insert(value);
            }
            else if (type == 2)
            {
                cout << "Character: ";
                cin >> character;
                chars.insert(character);
            }
            else
            {
                cout << "Word: ";
                cin >> word;
                strings.insert(word);
            }
        }
        else if (choice == 2)
        {
            if (type == 1)
            {
                cout << "Number: ";
                cin >> value;
                if(ints.search(value) != NULL)
                    cout << "The value " << value << " was found in the tree.\n";
                else
                    cout << "The value " << value << " was NOT found in the tree.\n";
            }
            else if (type == 2)
            {
                cout << "Character: ";
                cin >> character;
                if(chars.search(character) != NULL)
                    cout << "The character " << character << " was found in the tree.\n";
                else
                    cout << "The character " << character << " was NOT found in the tree.\n";
            }
            else
            {
                cout << "Word: ";
                cin >> word;
                if(strings.search(word) != NULL)
                    cout << "The word " << word << " was found in the tree.\n";
                else
                    cout << "The word " << word << " was NOT found in the tree.\n";
            }
        }
        else if (choice == 3)
        {
            cout << "Enter number of values:";
            cin >> size;
            if (type == 1)
            {
                for (int i = 0; i < size; i++)
                    ints.insert(rand() % 1000);
            }
            else if (type == 2)
            {
                for (int i = 0; i < size; i++)
                    chars.insert('a' + rand() % 26);
            }
            else
            {
                cout << "Word Lengths: ";
                cin >> length;
                for (int i = 0; i < size; i++)
                {
                    word.clear();
                    for (int i = 0; i < length; i++)
                        word += 'a' + rand() % 26;
                    strings.insert(word);
                }
            }
        }
        else if (choice == 4)
        {
            if (type == 1)
            {
                cout << "Number: ";
                cin >> value;
                ints.remove(value);
            }
            else if (type == 2)
            {
                cout << "Character: ";
                cin >> character;
                chars.remove(character);
            }
            else
            {
                cout << "Word: ";
                cin >> word;
                strings.remove(word);
            }
        }
        else if (choice == 5)
        {
             if (type == 1)
            {
                ints.print();
            }
            else if (type == 2)
            {
                chars.print();
            }
            else
            {
                strings.print();
            }
        }
    }
    return 0;

}


b_tree.cpp
#include "b_tree.h"

template <class dataType>
BTree<dataType>::BTree(int M)
{
    order = M;
    Root = NULL;
}

template <class dataType>
BTree<dataType>::BTree(const BTree& obj)
{
    order = obj.order;
    Root = obj.root;
}

template <class dataType>
BTree<dataType>::~BTree()
{
    //TODO add destructor
}

template <class dataType>
int BTree<dataType>::getOrder() const
{
    return order;
}

template <class dataType>
BTNode<dataType>* BTree<dataType>::getRoot() const
{
    return Root;
}

template <class dataType>
void BTree<dataType>::setRoot(BTNode<dataType>* root)
{
    Root = root;
}

template <class dataType>
BTNode<dataType>* BTree<dataType>::search(dataType val)
{
    return searchHelper(val, Root);
}

template <class dataType>
BTNode<dataType>* BTree<dataType>::searchHelper(dataType val, BTNode<dataType>* curr)
{
    //Terminating Condition 1: Node is NULL
    if(curr == NULL)
        return NULL;

    //TODO make a binary search function
    for(int i = 0; i < order - 1; i++)
    {
        dataType* arr = curr -> getDataArray();
        if(arr[i] == val)
            return curr;
        //Early Stopping
        else if(arr[i] > val)
            return searchHelper(val, curr -> getChildren[i]);
    }

}

template <class dataType>
void BTree<dataType>::insertHelper(dataType val, BTNode* currRoot)
{
    //Try to insert into root node by checking if space is avalible
    if(!currRoot -> isFull())
    {
        //Insert into array
        int i
        for(i = curr -> getDataArraySize() - 1; val < arr[i]; i--)
           arr[i + 1] = arr[i];
        arr[++i] = val;

        /*
            Test for Loop: M = 5, inserting: 3
            1 2 4
            i = curr -> getDataArraySize() - 1 = 3 - 1 = 2
            val = 3
            val = 3 < arr[i = 2] = 4 // Condition Met
            arr[i + 1 = 3] = arr[i = 2] // 1 2 4 4
            i--: i = 1
            val = 3 > arr[i = 1] = 2 // Condition Failed
            arr[++i: 2] = val // 1 2 3 4
        */

     }

        //Update dataArraySize
        curr -> incDataArraySize();
}


b_tree_node.h
#ifndef B_TREE_H
#define B_TREE_H

//-------------------------------------------------
// FileName: b_tree_node.h
// Purpose: Define and implement the methods and data associated with the BTNode class
// Date: 07/17/18
// Authors: Chris Carter, Davis Campbell, Wesley Nimmo, Sam Yeager
//-------------------------------------------------
//Using a template format for flexibility
template <class DataType>
class BTNode{
    public:
        //Constructors
        BTNode(int degree_, bool leaf_);

        // Deconstructor
        ~BTNode();

        //Methods
        bool isFull() const;
        void remove(DataType value_);

        template <class dataType>
        friend class BTree;

    private:
        int degree;
        int keys;
        bool leaf;
        DataType* dataArray;
        BTNode** children;
};

#include "b_tree_node.h"
#include <iostream>

using namespace std;

//-------------------------------------------------
// Method: Constructor
// Purpose: Initialize the private variables for the tree
// Date: 07/17/18
// Authors: Chris Carter, Davis Campbell, Wesley Nimmo, Sam Yeager
//-------------------------------------------------

template <class DataType>
BTNode<DataType>::BTNode(int degree_, bool leaf_)
{
    dataArray = new DataType[2 * degree_ - 1];
    children = new BTNode*[2 * degree_];
    degree = degree_;
    keys = 0;
    leaf = leaf_;
    for (int i = 0; i < 2 * degree_; i++)
        children[i] = NULL;
}

//-------------------------------------------------
// Method: Deconstructor
// Purpose: Free the memory associated with the BTNode
// Date: 07/17/18
// Author: Chris Carter
//-------------------------------------------------
template <class DataType>
BTNode<DataType>::~BTNode()
{
    delete[] dataArray;
    delete[] children;
}

//-------------------------------------------------
// Method: isFull
// Purpose: Return True if the node is full
// Date: 07/17/18
// Authors: Chris Carter, Davis Campbell
//-------------------------------------------------
template <class DataType>
bool BTNode<DataType>::isFull() const
{
    return keys == (2 * degree) - 1;
}

//-------------------------------------------------
// Method: remove
// Purpose: remove an element from the tree
// Date: 07/29/18
// Author: Davis Campbell
// Logic and general framework for algorithm from geeksforgeeks.org
//-------------------------------------------------
template <class DataType>
void BTNode<DataType>::remove(DataType value_)
{
    int i = 0;
    while(i < keys && dataArray[i] < value_)
         i++;

    if(dataArray[i] != value_ || i >= keys)
    {
        if(leaf)
        {
            cout << "The value " << value_ << " is not in this tree.\n";
            return;
        }

        bool keyPresent = (i == keys? true : false);

        // We have to make sure the child has enough keys for when we delete a key
        if (children[i]->keys < degree)
        {
            // From the logic, if the next child node has more than degree-1 keys we take a key from that child
            if (i != keys && children[i + 1]->keys > degree - 1)
            {
                BTNode* First = children[i];
                BTNode* Next = children[i + 1];

                // copy the value at the i to the end of the First node
                First->dataArray[First->keys] = dataArray[i];

                // copy the first child from the Next node to the end of the first node
                if(!First->leaf)
                    First->children[First->keys + 1] = Next->children[0];

                // copy the first value of the Next node to position i
                dataArray[i] = Next->dataArray[0];

                // Move all the values back in the Next node
                for (int j = 0; j < Next->keys; j++)
                    Next->dataArray[j] = Next->dataArray[j + 1];

                // move all the children pointers back in the Next node if it is not a leaf
                if (!Next->leaf)
                    for (int k = 0; k < Next->keys; k++)
                        Next->children[k] = Next->children[k + 1];

                // change the number of keys in each changed node
                Next->keys = Next->keys - 1;
                First->keys = First->keys + 1;
            }

            // From the logic, if the previous child node has more than degree - 1 keys, take a key from that child
            else if (i != 0 && children[i - 1]->keys > degree - 1)
            {
                BTNode* First = children[i];
                BTNode* Previous = children[i - 1];

                // move all the data in First to the right
                for (int j = First->keys - 1; j > -1; j--)
                    First->dataArray[j+1] = First->dataArray[j];

                //move all the children over as well
                if (!First->leaf)
                    for(int j = First->keys; j > -1; j--)
                        First->children[j+1] = First->children[j];

                // move the value at i to the first spot in First
                First->dataArray[0] = dataArray[i-1];

                // move the last value in previous to the spot at i
                dataArray[i] = Previous->dataArray[Previous->keys];

                // move the previous child over
                if (!First->leaf)
                    First->children[0] = Previous->children[Previous->keys];

                // change number of keys in each
                First->keys = First->keys + 1;
                Previous->keys = Previous->keys - 1;
            }

            // From the logic, as a last option we merge it with its sibling.
            else
            {
                // From the logic, if it's the last child node, merge it with its predecessor, otherwise merge it with its successor
                if (i != keys)
                {
                    // merge with next sibling
                    BTNode* First = children[i];
                    BTNode* Next = children[i + 1];

                    // copy value at current position to the end of first
                    First->dataArray[degree - 1] = dataArray[i];

                    // copy child pointers over from the next node to the end of the first node
                    if (!First->leaf)
                    {
                        for(int j = 0; j <= Next->keys; j++)
                            First->children[degree + j] = Next->children[j];
                    }

                    // copy the values from next to first
                    for(int j = 0; j < Next->keys; j++)
                        First->dataArray[degree + j] = Next->dataArray[j];

                    // fill in the gap in the current node
                    for (int j = i + 2; j <= keys; j++)
                        children[j - 1] = children[j];
                    for (int j = i + 1; j < keys; j++)
                        dataArray[j - 1] = dataArray[j];

                    // update key counts and free memory
                    keys -= 1;
                    First->keys = First->keys + Next->keys + 1;
                    delete Next;
                }
                else
                {
                    // merge with previous sibling
                    BTNode* Previous = children[i - 1];
                    BTNode* First = children[i];

                    // copy value at current position to the end of first
                    Previous->dataArray[degree - 1] = dataArray[i-1];

                    // copy child pointers over from the next node to the end of the first node
                    if (!First->leaf)
                    {
                        for(int j = 0; j <= First->keys; j++)
                            Previous->children[degree + j] = First->children[j];
                    }

                    // copy the values from next to first
                    for(int j = 0; j < First->keys; j++)
                        Previous->dataArray[degree + j] = First->dataArray[j];

                    // fill in the gap in the current node
                    for (int j = i + 1; j <= keys; j++)
                        children[j - 1] = children[j];
                    for (int j = i; j < keys; j++)
                        dataArray[j - 1] = dataArray[j];

                    // update key counts and free memory
                    keys -= 1;
                    Previous->keys = Previous->keys + First->keys + 1;
                    delete First;
                }
            }
        }

        if (keyPresent && i > keys)
            children[i - 1]->remove(value_);
        else
            children[i]->remove(value_);
    }
    else if(dataArray[i] == value_)
    {
        if (!leaf)
        {
            // by the logic, if the child at i has at least as many keys as the degree, we delete the predecessor of the key to be deleted and then put it in the key to be deleted's space
            if (children[i]->keys >= degree)
            {
                BTNode* Predecessor = children[i];
                while(!Predecessor->leaf)
                    Predecessor = Predecessor->children[Predecessor->keys];
                DataType Prior = Predecessor->dataArray[Predecessor->keys - 1];
                dataArray[i] = Prior;
                children[i]->remove(Prior);
            }
            // basically the opposite of above
            else if (children[i + 1]->keys >= degree)
            {
                BTNode* Successor = children[i + 1];
                while (!Successor->leaf)
                    Successor = Successor->children[0];
                DataType Post = Successor->dataArray[0];
                dataArray[i] = Post;
                children[i + 1]->remove(Post);
            }
            // if both the child and the predecessor dont have enough, merge the child into the predecessor along with the value, delete the value and the child.
            else
            {
                // merge with next sibling
                BTNode* First = children[i];
                BTNode* Next = children[i + 1];

                // copy value at current position to the end of first
                First->dataArray[degree - 1] = dataArray[i];

                // copy child pointers over from the next node to the end of the first node
                if (!First->leaf)
                {
                    for(int j = 0; j <= Next->keys; j++)
                        First->children[degree + j] = Next->children[j];
                }

                // copy the values from next to first
                for(int j = 0; j < Next->keys; j++)
                    First->dataArray[degree + j] = Next->dataArray[j];

                // fill in the gap in the current node
                for (int j = i + 2; j <= keys; j++)
                    children[j - 1] = children[j];
                for (int j = i + 1; j < keys; j++)
                    dataArray[j - 1] = dataArray[j];

                // update key counts and free memory
                keys -= 1;
                First->keys = First->keys + Next->keys + 1;
                delete Next;

                children[i]->remove(value_);
            }
        }
        else
        {
            // move keys backward and reduce count
            for (int j = i + 1; j < keys; j++)
                dataArray[j - 1] = dataArray[j];
            keys--;
        }
    }
}

#endif


b_tree_node.cpp
#include "b_tree_node.h"
#include <iostream>

using namespace std;

template <class DataType>
BTNode<DataType>::BTNode(int M = 4)
{
    dataArray = new DataType[M];
    children = new BTNode*[M];
    order = M;
}

template <class DataType>
BTNode<DataType>::BTNode(const BTNode &obj)
{
    //Implement a deep copy
}

template <class DataType>
BTNode<DataType>::~BTNode()
{
    delete[] dataArray;
    delete[] children;
}

template <class DataType>
DataType* BTNode<DataType>::getDataArray() const
{
    return dataArray;
}

template <class DataType>
BTNode<DataType>** BTNode<DataType>::getChildren() const
{
    return children;
}

template <class DataType>
void BTNode<DataType>::setDataArray(DataType* arr)
{
    for(int i = 0; i < order - 1; i++)
        dataArray[i] = arr[i];
}

template <class DataType>
void BTNode<DataType>::setChildren(BTNode** Children)
{
    children = Children;
}


int main()
{
    BTNode<int> test(4);

    //Setting the array

    const int size = 3;
    int array[size];

    for(int i = 0; i < size; i++)
        array[i] = i + 1;

    test.setDataArray(array);

    int* print;
    print = test.getDataArray();

    for(int i = 0; i < 3; i++)
        cout << print[i] << " -> ";

    cout << endl;

}

template <class DataType>
bool BTNode<DataType>::isEmpty() const
{
    if(dataArraySize == 0)
        return true;
    else
        return false;
}

template <class DataType>
bool BTNode<DataType>::isFull() const
{
    if(dataArraySize == order - 1)
        return true;
    else
        return false;
}


b_tree.h
//-------------------------------------------------
// FileName: b_tree.h
// Purpose: Define and implement the methods and data associated with the BTree class
// Date: 07/17/18
// Authors: Chris Carter, Davis Campbell, Wesley Nimmo, Sam Yeager
//-------------------------------------------------


#include "b_tree_node.h"
#include <cstdlib>

template <class dataType>
class BTree{

    public:
        BTree(int degree_);
        ~BTree();

        //Common Data Structures Methods
        BTNode<dataType>* search(dataType val);
        void insert(dataType value_);
        void remove(dataType value_);
        void print() const;

    private:
        //Member Variables
        int degree;
        BTNode<dataType>* Root;

        //Helper Functinos
        BTNode<dataType>* searchHelper(dataType val, BTNode<dataType>* curr);
        void printHelper(BTNode<dataType>* current) const;
	void deconstructHelper(BTNode<dataType>* current) const;
        void insertHelper(dataType value_, BTNode<dataType>* current);
};

//-------------------------------------------------
// Method: Constructor
// Purpose: Initialize the private variables for the tree
// Date: 07/17/18
// Author: Chris Carter
//-------------------------------------------------
template <class dataType>
BTree<dataType>::BTree(int degree_)
{
    degree = degree_;
    Root = NULL;
}

//-------------------------------------------------
// Method: Deconstructor
// Purpose: free space from the tree
// Date: 08/01/18
// Author: Wesley Nimmo
//-------------------------------------------------
template <class dataType>
BTree<dataType>::~BTree()
{
    if (Root != NULL)
        deconstructHelper(Root);
}

//-------------------------------------------------
// Method: Search
// Purpose: Search for an element within the tree
// Date: 07/17/18
// Author: Chris Carter
//-------------------------------------------------
template <class dataType>
BTNode<dataType>* BTree<dataType>::search(dataType val)
{
    return searchHelper(val, Root);
}

//-------------------------------------------------
// Method: SearchHelper
// Purpose: Help to search for an element in the tree
// Date: 07/17/18
// Author: Chris Carter
//-------------------------------------------------

template <class dataType>
BTNode<dataType>* BTree<dataType>::searchHelper(dataType val, BTNode<dataType>* curr)
{
    //Terminating Condition 1: Node is NULL
    if(curr == NULL)
        return NULL;

    // binary search function
    for(int i = 0; i < (2 * degree) - 1; i++)
    {
        if(curr->dataArray[i] == val)
            return curr;
        //Early Stopping
        else if(curr->dataArray[i] > val)
            return searchHelper(val, curr -> children[i]);
    }

}

//-------------------------------------------------
// Method: Insert
// Purpose: Insert an element into the B-Tree
// Date: 08/01/18
// Author: Davis Campbell
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::insert(dataType value_)
{
    // if the tree is empty just make a new node
    if (Root == NULL)
    {
        Root = new BTNode<dataType>(degree, true);
        Root->dataArray[0] = value_;
        Root->keys = 1;
    }
    // if the tree is not empty
    else
    {
        // if the root is not full, run insertHelper to simply add to it
        if (!(Root->isFull()))
            insertHelper(value_, Root);
        // if it is full, we need to split it
        else
        {
        // splitting the root
            // making 2 new nodes for it to be split into
            BTNode<dataType>* newRoot = new BTNode<dataType>(degree, false);
            BTNode<dataType>* sibling = new BTNode<dataType>(degree, Root->leaf);

            // Moving data from the old root to the new sibling
            for (int j = 0; j < degree - 1; j++)
                sibling->dataArray[j] = Root->dataArray[j + degree];
            if (!(Root->leaf))
            {
                for (int j = 0; j < degree; j++)
                    sibling->children[j] = Root->children[j + degree];
            }

            // making space in the new root for the old median key
            for (int j = newRoot->keys; j >= 1; j--)
                newRoot->children[j+1] = newRoot->children[j];
            for (int j = newRoot->keys - 1; j >= 0; j--)
                newRoot->dataArray[j + 1] = newRoot->dataArray[j];

            newRoot->dataArray[0] = Root->dataArray[degree-1];

            // setting pointers and updating keys
            newRoot->children[0] = Root;
            newRoot->children[1] = sibling;
            Root->keys = degree - 1;
            sibling->keys = degree - 1;
            newRoot->keys++;
            Root = newRoot;

            // find the correct point at which we should look to insert
            int i = 0;
            if (newRoot->dataArray[0] < value_)
                i++;

            // Insert into the now for sure non-full node
            insertHelper(value_, newRoot->children[i]);
        }

    }
}

//-------------------------------------------------
// Method: Insert Helper
// Purpose: Insert an element into the B-Tree
// Date: 08/01/18
// Author: Davis Campbell
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::insertHelper(dataType value_, BTNode<dataType>* current)
{
    // set an index to the last data point position of the current array
    int i = current->keys - 1;

    // if the current is a leaf, we just move over the values and put it in
    if (current->leaf)
    {
        // move everything over
        while (i >= 0 && current->dataArray[i] > value_)
        {
            current->dataArray[i+1] = current->dataArray[i];
            i--;
        }

        // put in the value and update the number of keys
        current->dataArray[i+1] = value_;
        current->keys++;
    }
    // if current is not a leaf
    else
    {
        // get the position at which the insert should be
        while (i >= 0 && current->dataArray[i] > value_)
            i--;

        // if the child in which we would like to put the new value is full, we need to split it and move the median value up to current
        if (current->children[i+1]->isFull())
        {
            // new index value for convenience
            int q = i + 1;

            // only create one new node this time because we already have the parent node
            BTNode<dataType>* sibling = new BTNode<dataType>(current->children[i + 1]->degree, current->children[i + 1]->leaf);


            // move the data from current to the new node
            for (int j = 0; j < degree-1; j++)
                sibling->dataArray[j] = current->children[i + 1]->dataArray[j + degree];
            if (!current->children[i + 1]->leaf)
            {
                for (int j = 0; j < degree; j++)
                    sibling->children[j] = current->children[i + 1]->children[j + degree];
            }


            // make space in current for the median node which is moving up
            for (int j = current->keys; j >= q+1; j--)
                current->children[j+1] = current->children[j];
            for (int j = current->keys - 1; j >= q; j--)
                current->dataArray[j+1] = current->dataArray[j];

            // put the node in place
            current->dataArray[q] = current->children[i + 1]->dataArray[degree - 1];

            // update pointers and keys
            current->children[i + 1]->keys = degree - 1;
            sibling->keys = degree - 1;
            current->keys++;
            current->children[q+1] = sibling;

            // see which child we would like to go into
            if (current->dataArray[i+1] < value_)
                i++;
        }
        // do it all again with that child!
        insertHelper(value_, current->children[i+1]);
    }
}

//-------------------------------------------------
// Method: Remove
// Purpose: Remove an element from the B Tree
// Date: 07/28/18
// Author: Davis Campbell
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::remove(dataType value_)
{
    if(Root == NULL)
    {
        cout << "Error: Can not delete from empty tree";
        return;
    }

    Root->remove(value_);

    // handle the case that the root is now empty
    if (Root->keys == 0)
    {
        BTNode<dataType>* temp = Root;
        if(Root->leaf)
            Root = NULL;
        else
            Root = Root->children[0];
        delete temp;
    }
}

//-------------------------------------------------
// Method: Print
// Purpose: Print the tree in sorted order
// Date: 08/01/18
// Author: Davis Campbell
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::print() const
{
    if (Root != NULL)
        printHelper(Root);
}

//-------------------------------------------------
// Method: Print Helper
// Purpose: print nodes recursively
// Date: 08/01/18
// Author: Davis Campbell
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::printHelper(BTNode<dataType>* current) const
{
    if (current != NULL)
    {
        // loop through the node (remember we have current->keys keys and current->keys + 1 children)
        int i;
        for (i = 0; i < current->keys; i++)
        {
            if(!current->leaf)
                printHelper(current->children[i]);
            cout << current->dataArray[i] << " ";
        }
        if(!current->leaf)
            printHelper(current->children[i]);
    }
}

//-------------------------------------------------
// Method: Deconstruct Helper
// Purpose: delete nodes recursively
// Date: 08/01/18
// Author: Wesley Nimmo
//-------------------------------------------------
template <class dataType>
void BTree<dataType>::deconstructHelper(BTNode<dataType>* current) const
{
    if (current != NULL)
    {
        // loop through the node (remember we have current->keys keys and current->keys + 1 children)
        int i;
        for (i = 0; i < current->keys; i++)
        {
            if(!current->leaf)
                deconstructHelper(current->children[i]);
            else
		delete current;
        }
        if(!current->leaf)
            deconstructHelper(current->children[i]);
    }
}
